### ISSUE TITLE
Fmriprep crashes when trying to use pediatric template on cluster with singularity

### ISSUE THREAD

**Issue by susanbenear** [nipreps/fmriprep#2717]
Hello everyone! I'm rather new to fmriprep, but have previously run it without issues on my university's high performance computing cluster using singularity, and specifying T1w space as well as the standard adult MNI template. However, when I try to run it using one of the pediatric atlases, I get an issue where, despite the fact that the template lives in my project directory on the cluster (I confirmed this), it seems fmriprep insists on trying to download it from the internet. This is unsuccessful each time, as our cluster does not have internet access. It is important to note that I did NOT have this issue when using the standard adult template or T1w space. 

I have previously posted about this issue on neurostars, where I include the error, my code, and some other information: https://neurostars.org/t/fmriprep-using-singularity-on-hpc-cant-find-pediatric-mni-template/21373

**Comment by susanbenear** [comment]
I'm using fmriprep 20.2.6!

**Comment by effigies** [comment]
Hi, can you share the latest command that you ran?

**Comment by susanbenear** [comment]
Here's the code I tried running on the cluster (exactly this code ran with no errors when I used the adult MNI template):

````
module load singularity

export 
SINGULARITYENV_TEMPLATEFLOW_HOME=/home/fmriprep/.cache/templateflow
                                 
rm -f archive/job_scripts/cmd_fmriprep_${PBS_JOBID}.txt
touch archive/job_scripts/cmd_fmriprep_${PBS_JOBID}.txt

for subj in ${subjects}; do
    echo singularity run --cleanenv \
            -B ~/work/project_dir/archive/templateflow:/opt/templateflow \
            ~/work/project_dir/archive/fmriprep-20.2.6.simg \
            ~/work/project_dir ~/work/project_dir/derivatives \
            participant --participant-label ${subj} \
            --output-spaces MNIPediatricAsym:res-2:cohort-2 \
            --fs-license-file ~/work/license.txt \
            --notrack \
            --stop-on-first-crash \
            --nthreads 8 --omp-nthreads 8 \
            --skip_bids_validation >> archive/job_scripts/cmd_fmriprep_${PBS_JOBID}.txt
done
                                                                         
torque-launch -p archive/job_scripts/chk_fmriprep_${PBS_JOBID}.txt 
archive/job_scripts/cmd_fmriprep_${PBS_JOBID}.txt
````

(FYI I run the command as a batch job, feeding it my list of subjects (as `${subjects}`) and instructing it to run in batches of 6)

**Comment by susanbenear** [comment]
This is the error message I get - fmriprep crashes after about half an hour without output.

````
pandoc-citeproc: reference mnipediatricasym not found
Downloading https://templateflow.s3.amazonaws.com/tpl-MNIPediatricAsym/cohort-2\
/tpl-MNIPediatricAsym_cohort-2_res-2_T1w.nii.gz
````
It can't download the template because our cluster doesn't have internet access, but it shouldn't need to download the template because it's in my templateflow directory!

**Comment by effigies** [comment]
Looks like you want to use this to make sure that the templateflow directory you're binding in is being used:

```
export SINGULARITYENV_TEMPLATEFLOW_HOME=/opt/templateflow
```

You can run `singularity shell` and check that it looks right from inside the container (`$` indicates normal bash shell, `>` indicates a shell inside the container):

```
$ export SINGULARITYENV_TEMPLATEFLOW_HOME=/opt/templateflow
$ singularity shell --cleanenv \
            -B ~/work/project_dir/archive/templateflow:/opt/templateflow \
            ~/work/project_dir/archive/fmriprep-20.2.6.simg
> ls /opt/templateflow
```

**Comment by susanbenear** [comment]
Yep, I ran this, and the template is indeed in the templateflow directory:

![Screen Shot 2022-02-17 at 4 35 45 PM](https://user-images.githubusercontent.com/54600997/154574705-bdfac81f-4b62-4828-8a84-2c1d541a15ff.png)

I also confirmed that the exact template I want exists in this directory and is not an empty file:

![Screen Shot 2022-02-17 at 4 38 16 PM](https://user-images.githubusercontent.com/54600997/154575011-44e5cea5-f173-42ce-b29e-4f2b6b3e16ca.png)

**Comment by susanbenear** [comment]
`learning_lemurs` is my `project_dir`, by the way. It's a kids study :)

**Comment by susanbenear** [comment]
Thanks so much for this recommendation! It seems like this might have worked for directing fmriprep to the pediatric MNI template, but now I get a lengthy error that seems to be another attempt to access the internet and find a different template (see screen shot). As far as I can tell, this template doesn't exist in the templateflow repository, and I'm not sure what it's needed for. Any ideas?

<img width="567" alt="Screen Shot 2022-02-17 at 7 10 08 PM" src="https://user-images.githubusercontent.com/54600997/154597972-73cf0f22-1bc5-4d16-bbda-d7404c5fa340.png">

**Comment by effigies** [comment]
You can use this script to fetch the templates we bundle with the Docker image: https://github.com/nipreps/fmriprep/blob/master/scripts/fetch_templates.py

**Comment by susanbenear** [comment]
Great! So I would download these templates and then move them into my `/opt/templateflow` directory on the cluster?

Sorry for my endless problems and naivety, but I'm running into this error when I try to run the `fetch_templates.py` script in a terminal on my local laptop:
````
File "fetch_templates.py", line 129, in <module>
    import templateflow.api as tf
ModuleNotFoundError: No module named 'templateflow'
````
Assuming I'm missing a package or library of some sort?

**Comment by susanbenear** [comment]
I was able to find a workaround for downloading the templates using the instructions on this page: https://www.templateflow.org/usage/archive/
I will report back on whether I can now successfully get fmriprep output on the cluster using a pediatric template!

**Comment by susanbenear** [comment]
@effigies I have ensured that all my templates are indeed in my templateflow directory and that it is binding correctly in the command in my script, yet I am still getting the error that fmriprep is trying to download the pediatric template from the internet:
<img width="570" alt="Screen Shot 2022-02-18 at 4 39 27 PM" src="https://user-images.githubusercontent.com/54600997/154765022-9758e47a-b6ad-4f9a-83f4-276c713ece8c.png">
.

**Comment by susanbenear** [comment]
@effigies No problem! Thanks again for your guidance. 

The result of the above commands tells me all the pediatric atlas files live in the location I was trying to bind to in my initial script, which includes the `.cache` directory, e.g. for the T1 template:
````
PosixPath('/home/tuh38197/.cache/templateflow/tpl-MNIPediatricAsym/cohort-2/tpl-MNIPediatricAsym_cohort-2_res-2_T1w.nii.gz')
````
So it seems I'm currently directing fmriprep to the `/opt` directory and should be directing it to the `/.cache` directory, which is what I was trying to do in my initial script by binding the two together, but clearly that didn't work either! 

I wonder if the script should instead read something like:
````
export 
SINGULARITYENV_TEMPLATEFLOW_HOME=/home/tuh38197/.cache/templateflow

for subj in ${subjects}; do
    echo singularity run --cleanenv \
            -B ~/work/project_dir/archive/templateflow:/home/tuh38197/.cache/templateflow \
            ~/work/project_dir/archive/fmriprep-20.2.6.simg \
...
```

**Comment by susanbenear** [comment]
@effigies I tried the above script and got the same error with fmriprep trying to download the template from the internet. Totally at a loss for what else to do and really need to get this data preprocessed ASAP. Any ideas or anyone else you'd recommend asking about this issue? Thanks in advance

**Comment by effigies** [comment]
Hi Susan. All I can think is that you're not exporting your directory correctly. It looks like the export is on a separate line from the definition. Once you have your variables properly exported, you should be able to open the container, and perform the fetches you need. Those should persist.

**Comment by susanbenear** [comment]
In my actual script, they are on the same line, I just double checked. I suppose I will keep messing around with binding different directories in different ways. Thanks for your help
